File: crm/app/internal/adapters/inbound/http/segment_handler.go
// Package http — Segment HTTP handlers.
// Auth: DEFERRED — LAN-trust only (per plan).
package http

import (
	"context"
	"encoding/json"
	"log"
	"net/http"

	"github.com/go-chi/chi/v5"

	"github.com/vantt/data-integration/crm/app/internal/domain"
)

// SegmentReader covers read operations for segments.
type SegmentReader interface {
	GetSegment(ctx context.Context, segmentID string) (*domain.Segment, error)
	ListSegments(ctx context.Context) ([]domain.Segment, error)
	ListMembers(ctx context.Context, segmentID string) ([]domain.SegmentMember, error)
}

// SegmentWriter covers create/update and manual member management.
type SegmentWriter interface {
	CreateSegment(ctx context.Context, seg *domain.Segment) error
	UpdateSegment(ctx context.Context, seg *domain.Segment) error
	AddMember(ctx context.Context, segmentID, partyID string) error
	RemoveMember(ctx context.Context, segmentID, partyID string) error
}

// SegmentRefresher triggers dynamic segment re-evaluation.
type SegmentRefresher interface {
	RefreshDynamicSegment(ctx context.Context, segmentID string) (int, error)
}

// SegmentHandler wires segment endpoints.
type SegmentHandler struct {
	reader    SegmentReader
	writer    SegmentWriter
	refresher SegmentRefresher
}

// NewSegmentHandler constructs the handler.
func NewSegmentHandler(r SegmentReader, w SegmentWriter, rf SegmentRefresher) *SegmentHandler {
	return &SegmentHandler{reader: r, writer: w, refresher: rf}
}

// RegisterRoutes mounts segment routes on the given chi router.
func (h *SegmentHandler) RegisterRoutes(r chi.Router) {
	r.Get("/api/segments", h.listSegments)
	r.Post("/api/segments", h.createSegment)
	r.Patch("/api/segments/{id}", h.updateSegment)
	r.Post("/api/segments/{id}/refresh", h.refreshSegment)
	r.Get("/api/segments/{id}/members", h.listMembers)
	r.Post("/api/segments/{id}/members", h.addMember)
	r.Delete("/api/segments/{id}/members/{partyId}", h.removeMember)
}

// GET /api/segments
func (h *SegmentHandler) listSegments(w http.ResponseWriter, r *http.Request) {
	segs, err := h.reader.ListSegments(r.Context())
	if err != nil {
		log.Printf("segment handler: list: %v", err)
		writeError(w, http.StatusInternalServerError, "failed to list segments")
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"segments": segs})
}

// POST /api/segments
func (h *SegmentHandler) createSegment(w http.ResponseWriter, r *http.Request) {
	var body struct {
		Name        string  `json:"name"`
		Description string  `json:"description"`
		IsDynamic   bool    `json:"is_dynamic"`
		Definition  string  `json:"definition"`
		OwnerUserID *string `json:"owner_user_id"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}
	seg := &domain.Segment{
		Name:        body.Name,
		Description: body.Description,
		IsDynamic:   body.IsDynamic,
		Definition:  body.Definition,
		OwnerUserID: body.OwnerUserID,
	}
	if err := h.writer.CreateSegment(r.Context(), seg); err != nil {
		if domain.IsValidationError(err) {
			writeError(w, http.StatusBadRequest, err.Error())
		} else {
			log.Printf("segment handler: create: %v", err)
			writeError(w, http.StatusInternalServerError, "operation failed")
		}
		return
	}
	writeJSON(w, http.StatusCreated, seg)
}

// PATCH /api/segments/{id}
func (h *SegmentHandler) updateSegment(w http.ResponseWriter, r *http.Request) {
	segID := chi.URLParam(r, "id")
	seg, err := h.reader.GetSegment(r.Context(), segID)
	if err != nil {
		log.Printf("segment handler: update get: %v", err)
		writeError(w, http.StatusInternalServerError, "operation failed")
		return
	}
	if seg == nil {
		writeError(w, http.StatusNotFound, "segment not found")
		return
	}

	var body struct {
		Name        *string `json:"name"`
		Description *string `json:"description"`
		IsDynamic   *bool   `json:"is_dynamic"`
		Definition  *string `json:"definition"`
		OwnerUserID *string `json:"owner_user_id"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}
	if body.Name != nil {
		seg.Name = *body.Name
	}
	if body.Description != nil {
		seg.Description = *body.Description
	}
	if body.IsDynamic != nil {
		seg.IsDynamic = *body.IsDynamic
	}
	if body.Definition != nil {
		seg.Definition = *body.Definition
	}
	if body.OwnerUserID != nil {
		seg.OwnerUserID = body.OwnerUserID
	}
	if err := h.writer.UpdateSegment(r.Context(), seg); err != nil {
		if domain.IsValidationError(err) {
			writeError(w, http.StatusBadRequest, err.Error())
		} else {
			log.Printf("segment handler: update %s: %v", segID, err)
			writeError(w, http.StatusInternalServerError, "operation failed")
		}
		return
	}
	writeJSON(w, http.StatusOK, seg)
}

// POST /api/segments/{id}/refresh
func (h *SegmentHandler) refreshSegment(w http.ResponseWriter, r *http.Request) {
	segID := chi.URLParam(r, "id")
	n, err := h.refresher.RefreshDynamicSegment(r.Context(), segID)
	if err != nil {
		if domain.IsValidationError(err) {
			writeError(w, http.StatusBadRequest, err.Error())
		} else {
			log.Printf("segment handler: refresh %s: %v", segID, err)
			writeError(w, http.StatusInternalServerError, "refresh failed")
		}
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"segment_id": segID, "members_refreshed": n})
}

// GET /api/segments/{id}/members
func (h *SegmentHandler) listMembers(w http.ResponseWriter, r *http.Request) {
	segID := chi.URLParam(r, "id")
	members, err := h.reader.ListMembers(r.Context(), segID)
	if err != nil {
		log.Printf("segment handler: list members %s: %v", segID, err)
		writeError(w, http.StatusInternalServerError, "failed to list members")
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"members": members})
}

// POST /api/segments/{id}/members   body: {"party_id":"..."}
func (h *SegmentHandler) addMember(w http.ResponseWriter, r *http.Request) {
	segID := chi.URLParam(r, "id")
	var body struct {
		PartyID string `json:"party_id"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil || body.PartyID == "" {
		writeError(w, http.StatusBadRequest, "party_id is required")
		return
	}
	if err := h.writer.AddMember(r.Context(), segID, body.PartyID); err != nil {
		log.Printf("segment handler: add member %s→%s: %v", segID, body.PartyID, err)
		writeError(w, http.StatusInternalServerError, "operation failed")
		return
	}
	writeJSON(w, http.StatusCreated, map[string]string{"segment_id": segID, "party_id": body.PartyID})
}

// DELETE /api/segments/{id}/members/{partyId}
func (h *SegmentHandler) removeMember(w http.ResponseWriter, r *http.Request) {
	segID := chi.URLParam(r, "id")
	partyID := chi.URLParam(r, "partyId")
	if err := h.writer.RemoveMember(r.Context(), segID, partyID); err != nil {
		log.Printf("segment handler: remove member %s→%s: %v", segID, partyID, err)
		writeError(w, http.StatusInternalServerError, "operation failed")
		return
	}
	writeJSON(w, http.StatusOK, map[string]string{"status": "removed"})
}

File: crm/src/adapters/inbound/http/segment_handler.py
"""Segment HTTP handlers — FastAPI APIRouter prefix=/api.

Auth: POST/PATCH/DELETE mutations require X-CRM-Token header (CRM_API_TOKEN env var).
GET endpoints are read-only and unauthenticated.
"""
from __future__ import annotations

import logging
from dataclasses import asdict
from typing import Any, Optional

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel

from adapters.inbound.http.auth_dependency import require_api_token

logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# Request schemas
# ---------------------------------------------------------------------------

class SegmentCreateBody(BaseModel):
    name: str
    description: Optional[str] = None
    is_dynamic: bool = False
    definition: Optional[Any] = None
    owner_user_id: Optional[str] = None


class SegmentUpdateBody(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    is_dynamic: Optional[bool] = None
    definition: Optional[Any] = None
    owner_user_id: Optional[str] = None


class AddMemberBody(BaseModel):
    party_id: str


# ---------------------------------------------------------------------------
# Router factory
# ---------------------------------------------------------------------------

def make_segment_router(
    segment_svc: Any,
) -> APIRouter:
    """Return an APIRouter wired to *segment_svc* (SegmentService instance).

    Keeps the handler thin: validation lives in the service layer.
    """
    router = APIRouter(prefix="/api")

    # ── GET /segments ────────────────────────────────────────────────────────

    @router.get("/segments")
    def list_segments():
        try:
            segs = segment_svc.list_segments()
        except Exception as exc:
            logger.exception("segment handler: list")
            raise HTTPException(status_code=500, detail="failed to list segments") from exc
        return {"segments": [asdict(s) for s in segs]}

    # ── POST /segments ───────────────────────────────────────────────────────

    @router.post("/segments", status_code=201, dependencies=[Depends(require_api_token)])
    def create_segment(body: SegmentCreateBody):
        try:
            seg = segment_svc.create_segment(body.model_dump())
        except ValueError as exc:
            raise HTTPException(status_code=400, detail=str(exc)) from exc
        except Exception as exc:
            logger.exception("segment handler: create")
            raise HTTPException(status_code=500, detail="operation failed") from exc
        return asdict(seg)

    # ── PATCH /segments/{id} ─────────────────────────────────────────────────

    @router.patch("/segments/{segment_id}", dependencies=[Depends(require_api_token)])
    def update_segment(segment_id: str, body: SegmentUpdateBody):
        patch = {k: v for k, v in body.model_dump().items() if v is not None}
        try:
            segment_svc.update_segment(segment_id, **patch)
        except ValueError as exc:
            msg = str(exc)
            status = 404 if "not found" in msg else 400
            raise HTTPException(status_code=status, detail=msg) from exc
        except Exception as exc:
            logger.exception("segment handler: update %s", segment_id)
            raise HTTPException(status_code=500, detail="operation failed") from exc
        return {"status": "ok", "segment_id": segment_id}

    # ── POST /segments/{id}/refresh ──────────────────────────────────────────

    @router.post("/segments/{segment_id}/refresh", dependencies=[Depends(require_api_token)])
    def refresh_segment(segment_id: str):
        try:
            n = segment_svc.refresh_dynamic_segment(segment_id)
        except ValueError as exc:
            raise HTTPException(status_code=400, detail=str(exc)) from exc
        except Exception as exc:
            logger.exception("segment handler: refresh %s", segment_id)
            raise HTTPException(status_code=500, detail="refresh failed") from exc
        return {"segment_id": segment_id, "members_refreshed": n}

    # ── GET /segments/{id}/members ───────────────────────────────────────────

    @router.get("/segments/{segment_id}/members")
    def list_members(segment_id: str):
        try:
            members = segment_svc.list_members(segment_id)
        except Exception as exc:
            logger.exception("segment handler: list members %s", segment_id)
            raise HTTPException(status_code=500, detail="failed to list members") from exc
        return {"members": [asdict(m) for m in members]}

    # ── POST /segments/{id}/members ──────────────────────────────────────────

    @router.post("/segments/{segment_id}/members", status_code=201, dependencies=[Depends(require_api_token)])
    def add_member(segment_id: str, body: AddMemberBody):
        try:
            segment_svc.add_member(segment_id, body.party_id)
        except Exception as exc:
            logger.exception("segment handler: add member %s→%s", segment_id, body.party_id)
            raise HTTPException(status_code=500, detail="operation failed") from exc
        return {"segment_id": segment_id, "party_id": body.party_id}

    # ── DELETE /segments/{id}/members/{party_id} ─────────────────────────────

    @router.delete("/segments/{segment_id}/members/{party_id}", dependencies=[Depends(require_api_token)])
    def remove_member(segment_id: str, party_id: str):
        try:
            segment_svc.remove_member(segment_id, party_id)
        except Exception as exc:
            logger.exception("segment handler: remove member %s→%s", segment_id, party_id)
            raise HTTPException(status_code=500, detail="operation failed") from exc
        return {"status": "removed"}

    return router
